<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<title>PLATO: PLatform for the Analysis, Translation, and Organization of large-scale data</title>
<link rel="icon" href="https://chgr.mc.vanderbilt.edu/plato/steps/favicon.ico" type="image/icon" />
<link rel="shortcut icon" href="https://chgr.mc.vanderbilt.edu/plato/steps/favicon.ico" type="image/icon" />
<link href="../stylesheets/cake.generic.css%3F1262635684.css" media="screen" rel="stylesheet" type="text/css" />
<link href="../stylesheets/wasp.css%3F1262635684.css" media="screen" rel="stylesheet" type="text/css" />
</head>
<body>
	<a name="top"></a>
	<div id="container">
		<div id="header">
			<h1>
<!--				<a href="/plato/"><img alt="Wasp_icon32x32" src="/plato/images/WASP_icon32x32.PNG?1262635684" /></a> -->
				<a href="../index.html">PLATO:</a> PLatform for the Analysis, Translation, and Organization of large-scale data
			</h1>
		</div>
		<div id="main" class="main">
			<?php //include("side-menu.php");?>
			<table border=0 cellpadding=0>
			<tr>
						<td id="side-menu" class="side-menu">
		       	<br/>
				<ul>
					<li><a href="../index.html">Home</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu/ritchielab/subscriptions">Download</a></li>
					<li><a href="../main/aboutus.html">About Us</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu/ritchielab/index.php">Ritchie Lab</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu">CHGR</a></li>
				</ul>
<br/>
<!-- Google CSE Search Box Begins  -->
<form method="GET" action="http://searchvu.vanderbilt.edu/search">
    <input type="text" name="q" maxlength="256" value="" />
    <input type="hidden" name="site" value="default_collection" />  
    <input type="hidden" name="client" value="default_frontend" />
    <input type="hidden" name="proxystylesheet" value="default_frontend" />
    <input type="hidden" name="output" value="xml_no_dtd" />   
    <input type="hidden" name="sitesearch" value="https://edam.mc.vanderbilt.edu/wasp" />
    <input type="submit" name="btnG" value="Search" />
</form>
<!-- Google CSE Search Box Ends -->

				<br/>
				<hr/>
				<br/>
<!--				<?php include("side-menu.php");?>-->
				
					<ul>
					
					<li><a href="../main/usage.html">Usage</a></li>
					
						<ul>
							
							<li><a href="../commandlines.html">Commandline Arguments</a></li>
							
							
							<li><a href="../step_options.html">Step Specific Arguments</a></li>
							
						</ul>
					</ul>
						<br/>
					<ul>
						<li>Input Files</li>
						<ul>
							
							<li><a href="../inputfiles/batchfiles.html">Batch Files</a></li>
							
							
							<li><a href="../inputfiles/pedfile.html">PED File</a></li>
							
							
							<li><a href="../inputfiles/mapfile.html">MAP File</a></li>
							
							
							<li><a href="../inputfiles/tpedfile.html">TPED File</a></li>
							
							
							<li><a href="../inputfiles/tfamfile.html">TFAM File</a></li>
							
							
							<li><a href="../inputfiles/binary.html">Binary</a></li>
							
							
							<li><a href="../inputfiles/covtraitfile.html">Covariates/Traits</a></li>
							

							
							<li><a href="../inputfiles/freqfile.html">Frequency File</a></li>
							
							
							<li><a href="../inputfiles/strandfile.html">Strand File</a></li>
							
							
							<li><a href="../inputfiles/mapfileext.html">Extended Map File</a></li>
							
							
							<li><a href="../inputfiles/sampfileext.html">Extended Sample File</a></li>
							
							
							<li><a href="../inputfiles/incexclists.html">Include/Exclude Files</a></li>
							
							
							<li><a href="../inputfiles/stratfile.html">Stratification File</a></li>
							
							
							<li><a href="../inputfiles/filterprocesscfg.html">Filter Process Config.</a></li>
							

						</ul>
					</ul>
						<br/>

					<ul>
						<li>QC Steps</li>
								<ul>
									
										<li>
										<a href="1.html">Marker Genotype Efficiency</a>
										</li>
									
										<li>
										<a href="2.html">Allele Frequency</a>
										</li>
									
										<li>
										<a href="3.html">Concordance</a>
										</li>
									
										<li>
										<a href="4.html">Gender Error</a>
										</li>
									
										<li>
										<a href="5.html">Hardy-Weinberg</a>
										</li>
									
										<li>
										<a href="6.html">Mendelian Errors</a>
										</li>
									
										<li>
										<a href="7.html">Mitochondrial Error</a>
										</li>
									
										<li>
										<a href="8.html">Sample Efficiency</a>
										</li>
									
										<li>
										<a href="33.html">Plato Filter Processing</a>
										</li>
									
								</ul>
						</ul>
						<br/>

						<ul>
							<li>Analysis Steps</li>
								<ul>
									
										<li>
											<a href="9.html">Chisquare</a>
										</li>
										
										<li>
											<a href="10.html">Logistic Regression</a>
										</li>
										
										<li>
											<a href="11.html">Linear Regression</a>
										</li>
										
										<li>
											<a href="12.html">Deletion Detection</a>
										</li>
										
										<li>
											<a href="13.html">Homozygous</a>
										</li>
										
										<li>
											<a href="14.html">Linkage Disequilibrium</a>
										</li>
										
										<li>
											<a href="15.html">TDT</a>
										</li>
										
										<li>
											<a href="30.html">IBS</a>
										</li>
										
										<li>
											<a href="31.html">CMH Analysis</a>
										</li>
										
										<li>
											<a href="32.html">Data Transformations</a>
										</li>
										
										<li>
											<a href="34.html">(EARTH) Multivariate adaptive regression splines</a>
										</li>
										
										<li>
											<a href="35.html">Kinship</a>
										</li>
										
								</ul>
						</ul>
						<br/>
						<ul>
							<li>Formatting Steps</li>
								<ul>
									
										<li>
											<a href="16.html">Beagle</a>
										</li>
										
										<li>
											<a href="17.html">FBAT</a>
										</li>
										
										<li>
											<a href="18.html">GRR</a>
										</li>
										
										<li>
											<a href="19.html">Lapis</a>
										</li>
										
										<li>
											<a href="20.html">MDR</a>
										</li>
										
										<li>
											<a href="21.html">PDT2</a>
										</li>
										
										<li>
											<a href="22.html">PED File</a>
										</li>
										
										<li>
											<a href="23.html">Phase</a>
										</li>
										
										<li>
											<a href="24.html">PowerMarker</a>
										</li>
										
										<li>
											<a href="25.html">QTDT</a>
										</li>
										
										<li>
											<a href="26.html">Structure</a>
										</li>
										
										<li>
											<a href="27.html">Superlink</a>
										</li>
										
										<li>
											<a href="28.html">TPED</a>
										</li>
										
										<li>
											<a href="29.html">Eigenstrat</a>
										</li>
										
										<li>
											<a href="36.html">Binary (.bim, .bed, .fam) Output</a>
										</li>
										
									
								</ul>
						</ul>
				<br/>

			</td>
			<td id="content" class="contents">
				
				<p style="color: green"></p>

				<h2>
  QTDT
</h2>



<p>
	Creates QTDT input files.
</p>

<br/>

<p>
	<a href="../inputfiles/batchfiles.html">Batch file</a> command:<br/>
	<pre class=code>
output-qtdt
	</pre>
</p>

<br/>

Step specific options:
<table class="datatable" style="background-color:#DBDBDB">

	<tr>
		<td>
			
				<b>
				-allele-custom {filename}			
				</b>
		</td>
		<td>
				Option for formatting steps to flip alleles based on custom mapping as specified in {filename}
				<br/>
				<br/>
				<b>Ex:</b>
				-allele-custom myallelemap.txt
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-allele12 			
				</b>
		</td>
		<td>
				Converts genotype output to 1s and 2s.  1 refers to the minor allele, 2 refers to the major allele.  Possible genotype codings include: 1 1, 1 2, 2 2, and 0 0 (for missing).
				<br/>
				<br/>
				<b>Ex:</b>
				output-ped -allele12
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-allele1234 			
				</b>
		</td>
		<td>
				Option for formatting steps to flip from ACGT to 1234 alleles
				<br/>
				<br/>
				<b>Ex:</b>
				-allele1234
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-alleleACGT 			
				</b>
		</td>
		<td>
				Option for formatting steps to flip from 1234 to ACGT
				<br/>
				<br/>
				<b>Ex:</b>
				-alleleACGT
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-bp-max {number}			
				</b>
		</td>
		<td>
				Specifies a maximum chromosome specific base-pair location to process. All markers with base-pairs greater than the number will be excluded for the specified step only. Can be used in conjunction with -bp-min and -bp-space. Note: -chrom is required.

				<br/>
				<br/>
				<b>Ex:</b>
				-chrom 3 -bp-max 15000000
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-bp-min {number}			
				</b>
		</td>
		<td>
				Specifies a minimum chromosome specific base-pair location to process. All markers with base-pairs less than the number will be excluded for the specified step only. Can be used in conjunction with -bp-max and -bp-space. Note: -chrom is required.

				<br/>
				<br/>
				<b>Ex:</b>
				-chrom 3 -bp-min 5000000
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-bp-space {number}			
				</b>
		</td>
		<td>
				Specifies to process only markers spaced by {number} base-pairs across all chromosomes. Can be used in conjunction with -bp-min and -bp-max. Note: -chrom is not required.

				<br/>
				<br/>
				<b>Ex:</b>
				-bp-space 1000000
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-center-file {filename}			
				</b>
		</td>
		<td>
				Specify a center code map file.<br/>
<br/>
File columns:<br/>
Family_ID<br/>
Sample_ID<br/>
Center_code
				<br/>
				<br/>
				<b>Ex:</b>
				-center-file mycenters.txt
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-chrom {number}			
				</b>
		</td>
		<td>
				Specifies a specific chromosome to process. Only this one chromosome will be processed for the step specified. (Note: X = 23, Y = 24, XY = 25, Mitochondrial = 26)
				<br/>
				<br/>
				<b>Ex:</b>
				-chrom 3
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-covars-name {list of columns}			
				</b>
		</td>
		<td>
				Specifies named columns to use in covariate file separated by commas.  Dashes are used to denote ranges of covariate columns.

				<br/>
				<br/>
				<b>Ex:</b>
				-covars-name AGE-NUM_FOUND,GRE,FN
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-covars-number {list of columns}			
				</b>
		</td>
		<td>
				Specifies column numbers in the covariate file to use separated by commas.  Dashes denote range of covariate columns.

				<br/>
				<br/>
				<b>Ex:</b>
				-covars-number 1-3,5,7
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-dummy-incomplete-parent-ids 			
				</b>
		</td>
		<td>
				This option will do one of two things when formatting output:<br/>
<br/>
In the case of a Sample having two parent ids in the PED file and one of those parent ids has no genotypes listed in the file, the resulting formatted output will add that missing sample to the file and fill in with 0/0 genotypes.<br/>
<br/>
In the case of a Sample having one parent id and the other one listed as missing (0), then the resulting output will fill in the 0 with DUM999999 for fathers and DUM999998 for mothers and add these dummy samples to the output filled in with 0/0 genotypes.
				<br/>
				<br/>
				<b>Ex:</b>
				output-ped -dummy-incomplete-parent-ids
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-inc-disabled-samples 			
				</b>
		</td>
		<td>
				Includes disabled samples in output of formatting steps to maintain pedigree information.

				<br/>
				<br/>
				<b>Ex:</b>
				-inc-disabled-samples
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-inc-excluded-samples 			
				</b>
		</td>
		<td>
				Include in output excluded samples specified by the command line argument(s) -incsamples or -excsamples

				<br/>
				<br/>
				<b>Ex:</b>
				-inc-excluded-samples
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-out {name}			
				</b>
		</td>
		<td>
				Overrides the suffix of the output files to be {name}
				<br/>
				<br/>
				<b>Ex:</b>
				-out test1
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-override-out {name}			
				</b>
		</td>
		<td>
				Overrides the default output filename(s) with whatever is specified in {name}.  Note: all filename extensions (.txt, .map, .ped, etc...) that Plato uses will be APPENDED to the name that is given.
				<br/>
				<br/>
				<b>Ex:</b>
				output-ped -override-out post_qc
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-traits-name {list of columns}			
				</b>
		</td>
		<td>
				Specifies named columns to use in trait file separated by commas.  Dashes denote range of trait columns.

				<br/>
				<br/>
				<b>Ex:</b>
				-traits-name AGE-NUM_FOUND,GRE,FN
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-traits-number {list of columns}			
				</b>
		</td>
		<td>
				Specifies column numbers in the trait file to use separated by commas.  Dashes denote range of trait columns to use.
				<br/>
				<br/>
				<b>Ex:</b>
				-traits-number 1-3,5,7
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-zero-disabled 			
				</b>
		</td>
		<td>
				Zero genotypes of disabled samples in formatting steps (-inc-disabled-samples required)

				<br/>
				<br/>
				<b>Ex:</b>
				-inc-disabled-samples -zero-disabled
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-zero-excluded 			
				</b>
		</td>
		<td>
				Zero genotypes of excluded samples specified by the command line argument(s) -incsamples or -excsamples (intended to be used with -inc-excluded-samples)

				<br/>
				<br/>
				<b>Ex:</b>
				-inc-excluded-samples -zero-exclude
		</td>
			
	</tr>

</table>
	


			</td>
			</tr>
			</table>
		</div>
		<div id="footer">
			&nbsp;
			
					<a href="../admin/login.html">Login</a> 
			
			&nbsp;
		</div>
	</div>
</body>
</html>