<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<title>PLATO: PLatform for the Analysis, Translation, and Organization of large-scale data</title>
<link rel="icon" href="http://localhost:3000/steps/favicon.ico" type="image/icon" />
<link rel="shortcut icon" href="http://localhost:3000/steps/favicon.ico" type="image/icon" />
<link href="../stylesheets/cake.generic.css%3F1248722767.css" media="screen" rel="stylesheet" type="text/css" />
<link href="../stylesheets/wasp.css%3F1243441516.css" media="screen" rel="stylesheet" type="text/css" />
</head>
<body>
	<a name="top"></a>
	<div id="container">
		<div id="header">
			<h1>
<!--				<a href="/"><img alt="Wasp_icon32x32" src="/images/WASP_icon32x32.PNG?1243448767" /></a> -->
				<a href="../index.html">PLATO:</a> PLatform for the Analysis, Translation, and Organization of large-scale data
			</h1>
		</div>
		<div id="main" class="main">
			<?php //include("side-menu.php");?>
			<table border=0 cellpadding=0>
			<tr>
						<td id="side-menu" class="side-menu">
		       	<br/>
				<ul>
					<li><a href="../index.html">Home</a></li>
					<li><a href="../downloads.html">Download</a></li>
					<li><a href="../main/aboutus.html">About Us</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu">CHGR</a></li>
				</ul>
<br/>
<!-- Google CSE Search Box Begins  -->
<form method="GET" action="http://searchvu.vanderbilt.edu/search">
    <input type="text" name="q" maxlength="256" value="" />
    <input type="hidden" name="site" value="default_collection" />  
    <input type="hidden" name="client" value="default_frontend" />
    <input type="hidden" name="proxystylesheet" value="default_frontend" />
    <input type="hidden" name="output" value="xml_no_dtd" />   
    <input type="hidden" name="sitesearch" value="https://edam.mc.vanderbilt.edu/wasp" />
    <input type="submit" name="btnG" value="Search" />
</form>
<!-- Google CSE Search Box Ends -->

				<br/>
				<hr/>
				<br/>
<!--				<?php include("side-menu.php");?>-->
				
					<ul>
					
					<li><a href="../main/usage.html">Usage</a></li>
					
						<ul>
							
							<li><a href="../commandlines.html">Commandline Arguments</a></li>
							
							
							<li><a href="../step_options.html">Step Specific Arguments</a></li>
							
						</ul>
					</ul>
						<br/>
					<ul>
						<li>Input Files</li>
						<ul>
							
							<li><a href="../inputfiles/batchfiles.html">Batch Files</a></li>
							
							
							<li><a href="../inputfiles/pedfile.html">PED File</a></li>
							
							
							<li><a href="../inputfiles/mapfile.html">MAP File</a></li>
							
							
							<li><a href="../inputfiles/tpedfile.html">TPED File</a></li>
							
							
							<li><a href="../inputfiles/tfamfile.html">TFAM File</a></li>
							
							
							<li><a href="../inputfiles/binary.html">Binary</a></li>
							
							
							<li><a href="../inputfiles/covtraitfile.html">Covariates/Traits</a></li>
							

							
							<li><a href="../inputfiles/freqfile.html">Frequency File</a></li>
							
							
							<li><a href="../inputfiles/strandfile.html">Strand File</a></li>
							
							
							<li><a href="../inputfiles/mapfileext.html">Extended Map File</a></li>
							
							
							<li><a href="../inputfiles/sampfileext.html">Extended Sample File</a></li>
							
							
							<li><a href="../inputfiles/incexclists.html">Include/Exclude Files</a></li>
							
							
							<li><a href="../inputfiles/stratfile.html">Stratification File</a></li>
							
							
							<li><a href="../inputfiles/filterprocesscfg.html">Filter Process Config.</a></li>
							

						</ul>
					</ul>
						<br/>

					<ul>
						<li>QC Steps</li>
								<ul>
									
										<li>
										<a href="1.html">Marker Genotype Efficiency</a>
										</li>
									
										<li>
										<a href="2.html">Allele Frequency</a>
										</li>
									
										<li>
										<a href="3.html">Concordance</a>
										</li>
									
										<li>
										<a href="4.html">Gender Error</a>
										</li>
									
										<li>
										<a href="5.html">Hardy-Weinberg</a>
										</li>
									
										<li>
										<a href="6.html">Mendelian Errors</a>
										</li>
									
										<li>
										<a href="7.html">Mitochondrial Error</a>
										</li>
									
										<li>
										<a href="8.html">Sample Efficiency</a>
										</li>
									
										<li>
										<a href="33.html">Plato Filter Processing</a>
										</li>
									
								</ul>
						</ul>
						<br/>

						<ul>
							<li>Analysis Steps</li>
								<ul>
									
										<li>
											<a href="9.html">Chisquare</a>
										</li>
										
										<li>
											<a href="10.html">Logistic Regression</a>
										</li>
										
										<li>
											<a href="11.html">Linear Regression</a>
										</li>
										
										<li>
											<a href="12.html">Deletion Detection</a>
										</li>
										
										<li>
											<a href="13.html">Homozygous</a>
										</li>
										
										<li>
											<a href="14.html">Linkage Disequilibrium</a>
										</li>
										
										<li>
											<a href="15.html">TDT</a>
										</li>
										
										<li>
											<a href="30.html">IBS</a>
										</li>
										
										<li>
											<a href="31.html">CMH Analysis</a>
										</li>
										
										<li>
											<a href="32.html">Data Transformations</a>
										</li>
										
										<li>
											<a href="34.html">(EARTH) Multivariate adaptive regression splines</a>
										</li>
										
										<li>
											<a href="35.html">Kinship</a>
										</li>
										
								</ul>
						</ul>
						<br/>
						<ul>
							<li>Formatting Steps</li>
								<ul>
									
										<li>
											<a href="16.html">Beagle</a>
										</li>
										
										<li>
											<a href="17.html">FBAT</a>
										</li>
										
										<li>
											<a href="18.html">GRR</a>
										</li>
										
										<li>
											<a href="19.html">Lapis</a>
										</li>
										
										<li>
											<a href="20.html">MDR</a>
										</li>
										
										<li>
											<a href="21.html">PDT2</a>
										</li>
										
										<li>
											<a href="22.html">PED File</a>
										</li>
										
										<li>
											<a href="23.html">Phase</a>
										</li>
										
										<li>
											<a href="24.html">PowerMarker</a>
										</li>
										
										<li>
											<a href="25.html">QTDT</a>
										</li>
										
										<li>
											<a href="26.html">Structure</a>
										</li>
										
										<li>
											<a href="27.html">Superlink</a>
										</li>
										
										<li>
											<a href="28.html">TPED</a>
										</li>
										
										<li>
											<a href="29.html">Eigenstrat</a>
										</li>
										
									
								</ul>
						</ul>
				<br/>

			</td>
			<td id="content" class="contents">
				
				<p style="color: green"></p>

				<h2>
  Mendelian Errors
</h2>



<p>
	The mendelian error step has two stages.<br/><br/>

Stage 1: This looks at each mother, father, child goup and performs simple checks for mendelian error. Each error encountered by an individual increments the individual error count. Each family incurs only one error increment per marker. Each marker only incurs one error increment per family. <br/><br/>

Stage 2: This stage emulates the level 2 error checking as seen in Pedcheck. The errors that occur here are displayed in a separate output file and are not added to the counts generated in stage 1. This is because some of the level 2 errors could be speculative as opposed to real error. 
</p>

<br/>

<p>
	<a href="../inputfiles/batchfiles.html">Batch file</a> command:<br/>
	<pre class=code>
mendelian-error
	</pre>
</p>

<br/>

Step specific options:
<table class="datatable" style="background-color:#DBDBDB">

	<tr>
		<td>
			
				<b>
				-bp-max {number}			
				</b>
		</td>
		<td>
				Specifies a maximum chromosome specific base-pair location to process. All markers with base-pairs greater than the number will be excluded for the specified step only. Can be used in conjunction with -bp-min and -bp-space. Note: -chrom is required.

				<br/>
				<br/>
				<b>Ex:</b>
				-chrom 3 -bp-max 15000000
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-bp-min {number}			
				</b>
		</td>
		<td>
				Specifies a minimum chromosome specific base-pair location to process. All markers with base-pairs less than the number will be excluded for the specified step only. Can be used in conjunction with -bp-max and -bp-space. Note: -chrom is required.

				<br/>
				<br/>
				<b>Ex:</b>
				-chrom 3 -bp-min 5000000
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-bp-space {number}			
				</b>
		</td>
		<td>
				Specifies to process only markers spaced by {number} base-pairs across all chromosomes. Can be used in conjunction with -bp-min and -bp-max. Note: -chrom is not required.

				<br/>
				<br/>
				<b>Ex:</b>
				-bp-space 1000000
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-out {name}			
				</b>
		</td>
		<td>
				Overrides the suffix of the output files to be {name}
				<br/>
				<br/>
				<b>Ex:</b>
				-out test1
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-thresh-families-max {number}			
				</b>
		</td>
		<td>
				Specifies an upper cutoff threshold for families. All families with values greater than the specified value will be removed from future steps.

				<br/>
				<br/>
				<b>Ex:</b>
				-thresh-families-max 100
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-thresh-families-min {number}			
				</b>
		</td>
		<td>
				Specifies a lower cutoff threshold for families. All families with values less than the specified value will be removed from future steps.

				<br/>
				<br/>
				<b>Ex:</b>
				-thresh-families-min 10
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-thresh-markers-max {number}			
				</b>
		</td>
		<td>
				Specifies an upper cutoff threshold for markers. All markers with values greater than the specified value will be removed from future steps.

				<br/>
				<br/>
				<b>Ex:</b>
				-thresh-markers-max 0.95
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-thresh-markers-min {number}			
				</b>
		</td>
		<td>
				Specifies a lower cutoff threshold for markers. All markers with values less than the specified value will be removed from future steps.

				<br/>
				<br/>
				<b>Ex:</b>
				-thresh-markers-min 0.95
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-thresh-samples-max {number}			
				</b>
		</td>
		<td>
				Specifies an upper cutoff threshold for samples. All samples with values greater than the specified value will be removed from future steps.

				<br/>
				<br/>
				<b>Ex:</b>
				-thresh-samples-max 0.95
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-thresh-samples-min {number}			
				</b>
		</td>
		<td>
				Specifies a lower cutoff threshold for samples. All samples with values less than the specified value will be removed from future steps.

				<br/>
				<br/>
				<b>Ex:</b>
				-thresh-samples-min 0.95
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-zero 			
				</b>
		</td>
		<td>
				Zeros out genotypes with mendelian errors.

				<br/>
				<br/>
				<b>Ex:</b>
				-zero
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-zero-l2 			
				</b>
		</td>
		<td>
				Zeros genotypes of all samples with level 2 genotype errors for the markers where the errors exist

				<br/>
				<br/>
				<b>Ex:</b>
				-zero-l2
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-zero-l2-fams 			
				</b>
		</td>
		<td>
				Zero genotypes of all samples in a family for level 2 mendelian errors
				<br/>
				<br/>
				<b>Ex:</b>
				-zero-l2-fams
		</td>
			
	</tr>

</table>
	


			</td>
			</tr>
			</table>
		</div>
		<div id="footer">
			&nbsp;
			
					<a href="../admin/login.html">Login</a> 
			
			&nbsp;
		</div>
	</div>
</body>
</html>