<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<title>PLATO: PLatform for the Analysis, Translation, and Organization of large-scale data</title>
<link rel="icon" href="http://localhost:3000/steps/favicon.ico" type="image/icon" />
<link rel="shortcut icon" href="http://localhost:3000/steps/favicon.ico" type="image/icon" />
<link href="../stylesheets/cake.generic.css%3F1248722767.css" media="screen" rel="stylesheet" type="text/css" />
<link href="../stylesheets/wasp.css%3F1243441516.css" media="screen" rel="stylesheet" type="text/css" />
</head>
<body>
	<a name="top"></a>
	<div id="container">
		<div id="header">
			<h1>
<!--				<a href="/"><img alt="Wasp_icon32x32" src="/images/WASP_icon32x32.PNG?1243448767" /></a> -->
				<a href="../index.html">PLATO:</a> PLatform for the Analysis, Translation, and Organization of large-scale data
			</h1>
		</div>
		<div id="main" class="main">
			<?php //include("side-menu.php");?>
			<table border=0 cellpadding=0>
			<tr>
						<td id="side-menu" class="side-menu">
		       	<br/>
				<ul>
					<li><a href="../index.html">Home</a></li>
					<li><a href="../downloads.html">Download</a></li>
					<li><a href="../main/aboutus.html">About Us</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu">CHGR</a></li>
				</ul>
<br/>
<!-- Google CSE Search Box Begins  -->
<form method="GET" action="http://searchvu.vanderbilt.edu/search">
    <input type="text" name="q" maxlength="256" value="" />
    <input type="hidden" name="site" value="default_collection" />  
    <input type="hidden" name="client" value="default_frontend" />
    <input type="hidden" name="proxystylesheet" value="default_frontend" />
    <input type="hidden" name="output" value="xml_no_dtd" />   
    <input type="hidden" name="sitesearch" value="https://edam.mc.vanderbilt.edu/wasp" />
    <input type="submit" name="btnG" value="Search" />
</form>
<!-- Google CSE Search Box Ends -->

				<br/>
				<hr/>
				<br/>
<!--				<?php include("side-menu.php");?>-->
				
					<ul>
					
					<li><a href="../main/usage.html">Usage</a></li>
					
						<ul>
							
							<li><a href="../commandlines.html">Commandline Arguments</a></li>
							
							
							<li><a href="../step_options.html">Step Specific Arguments</a></li>
							
						</ul>
					</ul>
						<br/>
					<ul>
						<li>Input Files</li>
						<ul>
							
							<li><a href="../inputfiles/batchfiles.html">Batch Files</a></li>
							
							
							<li><a href="../inputfiles/pedfile.html">PED File</a></li>
							
							
							<li><a href="../inputfiles/mapfile.html">MAP File</a></li>
							
							
							<li><a href="../inputfiles/tpedfile.html">TPED File</a></li>
							
							
							<li><a href="../inputfiles/tfamfile.html">TFAM File</a></li>
							
							
							<li><a href="../inputfiles/binary.html">Binary</a></li>
							
							
							<li><a href="../inputfiles/covtraitfile.html">Covariates/Traits</a></li>
							

							
							<li><a href="../inputfiles/freqfile.html">Frequency File</a></li>
							
							
							<li><a href="../inputfiles/strandfile.html">Strand File</a></li>
							
							
							<li><a href="../inputfiles/mapfileext.html">Extended Map File</a></li>
							
							
							<li><a href="../inputfiles/sampfileext.html">Extended Sample File</a></li>
							
							
							<li><a href="../inputfiles/incexclists.html">Include/Exclude Files</a></li>
							
							
							<li><a href="../inputfiles/stratfile.html">Stratification File</a></li>
							
							
							<li><a href="../inputfiles/filterprocesscfg.html">Filter Process Config.</a></li>
							

						</ul>
					</ul>
						<br/>

					<ul>
						<li>QC Steps</li>
								<ul>
									
										<li>
										<a href="1.html">Marker Genotype Efficiency</a>
										</li>
									
										<li>
										<a href="2.html">Allele Frequency</a>
										</li>
									
										<li>
										<a href="3.html">Concordance</a>
										</li>
									
										<li>
										<a href="4.html">Gender Error</a>
										</li>
									
										<li>
										<a href="5.html">Hardy-Weinberg</a>
										</li>
									
										<li>
										<a href="6.html">Mendelian Errors</a>
										</li>
									
										<li>
										<a href="7.html">Mitochondrial Error</a>
										</li>
									
										<li>
										<a href="8.html">Sample Efficiency</a>
										</li>
									
										<li>
										<a href="33.html">Plato Filter Processing</a>
										</li>
									
								</ul>
						</ul>
						<br/>

						<ul>
							<li>Analysis Steps</li>
								<ul>
									
										<li>
											<a href="9.html">Chisquare</a>
										</li>
										
										<li>
											<a href="10.html">Logistic Regression</a>
										</li>
										
										<li>
											<a href="11.html">Linear Regression</a>
										</li>
										
										<li>
											<a href="12.html">Deletion Detection</a>
										</li>
										
										<li>
											<a href="13.html">Homozygous</a>
										</li>
										
										<li>
											<a href="14.html">Linkage Disequilibrium</a>
										</li>
										
										<li>
											<a href="15.html">TDT</a>
										</li>
										
										<li>
											<a href="30.html">IBS</a>
										</li>
										
										<li>
											<a href="31.html">CMH Analysis</a>
										</li>
										
										<li>
											<a href="32.html">Data Transformations</a>
										</li>
										
										<li>
											<a href="34.html">(EARTH) Multivariate adaptive regression splines</a>
										</li>
										
										<li>
											<a href="35.html">Kinship</a>
										</li>
										
								</ul>
						</ul>
						<br/>
						<ul>
							<li>Formatting Steps</li>
								<ul>
									
										<li>
											<a href="16.html">Beagle</a>
										</li>
										
										<li>
											<a href="17.html">FBAT</a>
										</li>
										
										<li>
											<a href="18.html">GRR</a>
										</li>
										
										<li>
											<a href="19.html">Lapis</a>
										</li>
										
										<li>
											<a href="20.html">MDR</a>
										</li>
										
										<li>
											<a href="21.html">PDT2</a>
										</li>
										
										<li>
											<a href="22.html">PED File</a>
										</li>
										
										<li>
											<a href="23.html">Phase</a>
										</li>
										
										<li>
											<a href="24.html">PowerMarker</a>
										</li>
										
										<li>
											<a href="25.html">QTDT</a>
										</li>
										
										<li>
											<a href="26.html">Structure</a>
										</li>
										
										<li>
											<a href="27.html">Superlink</a>
										</li>
										
										<li>
											<a href="28.html">TPED</a>
										</li>
										
										<li>
											<a href="29.html">Eigenstrat</a>
										</li>
										
									
								</ul>
						</ul>
				<br/>

			</td>
			<td id="content" class="contents">
				
				<p style="color: green"></p>

				<h2>
  Allele Frequency
</h2>



<p>
	<p>By default, the overall allele frequency calculation is done using founders only. If no founders in a family have genotypes at a given marker, then a random genotyped individual is chosen. You can override how the overall calculation is done by using some of the options listed below. Micro-satellite variations are included in the allele frequency output file(s), however they are included, but don't have any values in the genotypic frequency files.
</p>
<br/>
<p>
Any thresholds that are specified are performed based on the overall calculation.</p>
</p>

<br/>

<p>
	<a href="../inputfiles/batchfiles.html">Batch file</a> command:<br/>
	<pre class=code>
allele-freq
	</pre>
</p>

<br/>

Step specific options:
<table class="datatable" style="background-color:#DBDBDB">

	<tr>
		<td>
			
				<b>
				-all 			
				</b>
		</td>
		<td>
				Override allele frequency overall calculation to use all samples. Cannot be used with -unaff-spouses-only, -unk-spouses, -random-child, or -all-children.
				<br/>
				<br/>
				<b>Ex:</b>
				-all
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-all-children 			
				</b>
		</td>
		<td>
				Override allele frequency overall calculation to use all children. Cannot be used with -unaff-spouses-only, -unk-spouses, -random-child, or -all.
				<br/>
				<br/>
				<b>Ex:</b>
				-all-children
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-bp-max {number}			
				</b>
		</td>
		<td>
				Specifies a maximum chromosome specific base-pair location to process. All markers with base-pairs greater than the number will be excluded for the specified step only. Can be used in conjunction with -bp-min and -bp-space. Note: -chrom is required.

				<br/>
				<br/>
				<b>Ex:</b>
				-chrom 3 -bp-max 15000000
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-bp-min {number}			
				</b>
		</td>
		<td>
				Specifies a minimum chromosome specific base-pair location to process. All markers with base-pairs less than the number will be excluded for the specified step only. Can be used in conjunction with -bp-max and -bp-space. Note: -chrom is required.

				<br/>
				<br/>
				<b>Ex:</b>
				-chrom 3 -bp-min 5000000
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-bp-space {number}			
				</b>
		</td>
		<td>
				Specifies to process only markers spaced by {number} base-pairs across all chromosomes. Can be used in conjunction with -bp-min and -bp-max. Note: -chrom is not required.

				<br/>
				<br/>
				<b>Ex:</b>
				-bp-space 1000000
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-casecontrol 			
				</b>
		</td>
		<td>
				Provide output borken down by male/female Case/Control.

				<br/>
				<br/>
				<b>Ex:</b>
				-casecontrol
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-chrom {number}			
				</b>
		</td>
		<td>
				Specifies a specific chromosome to process. Only this one chromosome will be processed for the step specified. (Note: X = 23, Y = 24, Mitochondrial = 25, XY = 26)
				<br/>
				<br/>
				<b>Ex:</b>
				-chrom 3
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-filter-file 			
				</b>
		</td>
		<td>
				Filter markers by minor allele frequencies as found in a predefined frequency file using the -freq-file {filename} command line argument.

				<br/>
				<br/>
				<b>Ex:</b>
				-filter-file
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-filter-overall 			
				</b>
		</td>
		<td>
				Filter markers by overall minor allele frequency.

				<br/>
				<br/>
				<b>Ex:</b>
				-thresh-markers-min 0.1 -filter-overall
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-gender 			
				</b>
		</td>
		<td>
				Provide output borken down by overall male/female.

				<br/>
				<br/>
				<b>Ex:</b>
				-gender
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-group-file {filename}			
				</b>
		</td>
		<td>
				Specifies a sample grouping (stratification) file to use. 
<br/>
<br/>
Columns:<br/>
FamID<br/>
IndID<br/>
Group
				<br/>
				<br/>
				<b>Ex:</b>
				-group-file mygroupfile.txt
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-out {name}			
				</b>
		</td>
		<td>
				Overrides the suffix of the output files to be {name}
				<br/>
				<br/>
				<b>Ex:</b>
				-out test1
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-parental 			
				</b>
		</td>
		<td>
				Provide output broken down by male/female parents.
				<br/>
				<br/>
				<b>Ex:</b>
				-parental
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-random-child 			
				</b>
		</td>
		<td>
				Override allele frequency overall calculation to use one random child from each family. Cannot be used with -unaff-spouses-only, -unk-spouses, -all-children, or -all.

				<br/>
				<br/>
				<b>Ex:</b>
				-random-child
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-rm-het-only 			
				</b>
		</td>
		<td>
				Removes all snps that only have heterozygous calls
				<br/>
				<br/>
				<b>Ex:</b>
				-rm-het-only
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-rm-mono 			
				</b>
		</td>
		<td>
				Filters all monomorphic markers based on the genotype data.

				<br/>
				<br/>
				<b>Ex:</b>
				-rm-mono
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-thresh-markers-max {number}			
				</b>
		</td>
		<td>
				Specifies an upper cutoff threshold for markers. All markers with values greater than the specified value will be removed from future steps.

				<br/>
				<br/>
				<b>Ex:</b>
				-thresh-markers-max 0.95
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-thresh-markers-min {number}			
				</b>
		</td>
		<td>
				Specifies a lower cutoff threshold for markers. All markers with values less than the specified value will be removed from future steps.

				<br/>
				<br/>
				<b>Ex:</b>
				-thresh-markers-min 0.95
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-unaff-spouses-only 			
				</b>
		</td>
		<td>
				Override allele frequency overall calculation to use unaffected spouses only. Cannot be used with -unk-spouses, -all-children, -random-child, or -all.

				<br/>
				<br/>
				<b>Ex:</b>
				-unaff-spouses-only
		</td>
			
	</tr>

	<tr>
		<td>
			
				<b>
				-unk-spouses 			
				</b>
		</td>
		<td>
				Override allele frequency overall calculation to use unknown spouses only. Cannot be used with -unaff-spouses-only, -all-children, -random-child, or -all.

				<br/>
				<br/>
				<b>Ex:</b>
				-unk-spouses
		</td>
			
	</tr>

</table>
	


			</td>
			</tr>
			</table>
		</div>
		<div id="footer">
			&nbsp;
			
					<a href="../admin/login.html">Login</a> 
			
			&nbsp;
		</div>
	</div>
</body>
</html>