<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<title>PLATO: PLatform for the Analysis, Translation, and Organization of large-scale data</title>
<link rel="icon" href="favicon.ico" type="image/icon" />
<link rel="shortcut icon" href="favicon.ico" type="image/icon" />
<link href="stylesheets/cake.generic.css%3F1262635684.css" media="screen" rel="stylesheet" type="text/css" />
<link href="stylesheets/wasp.css%3F1262635684.css" media="screen" rel="stylesheet" type="text/css" />
</head>
<body>
	<a name="top"></a>
	<div id="container">
		<div id="header">
			<h1>
<!--				<a href="/plato/"><img alt="Wasp_icon32x32" src="/plato/images/WASP_icon32x32.PNG?1262635684" /></a> -->
				<a href="index.html">PLATO:</a> PLatform for the Analysis, Translation, and Organization of large-scale data
			</h1>
		</div>
		<div id="main" class="main">
			<?php //include("side-menu.php");?>
			<table border=0 cellpadding=0>
			<tr>
						<td id="side-menu" class="side-menu">
		       	<br/>
				<ul>
					<li><a href="index.html">Home</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu/ritchielab/subscriptions">Download</a></li>
					<li><a href="main/aboutus.html">About Us</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu/ritchielab/index.php">Ritchie Lab</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu">CHGR</a></li>
				</ul>
<br/>
<!-- Google CSE Search Box Begins  -->
<form method="GET" action="http://searchvu.vanderbilt.edu/search">
    <input type="text" name="q" maxlength="256" value="" />
    <input type="hidden" name="site" value="default_collection" />  
    <input type="hidden" name="client" value="default_frontend" />
    <input type="hidden" name="proxystylesheet" value="default_frontend" />
    <input type="hidden" name="output" value="xml_no_dtd" />   
    <input type="hidden" name="sitesearch" value="https://edam.mc.vanderbilt.edu/wasp" />
    <input type="submit" name="btnG" value="Search" />
</form>
<!-- Google CSE Search Box Ends -->

				<br/>
				<hr/>
				<br/>
<!--				<?php include("side-menu.php");?>-->
				
					<ul>
					
					<li><a href="main/usage.html">Usage</a></li>
					
						<ul>
							
							<li><a href="commandlines.html">Commandline Arguments</a></li>
							
							
							<li><a href="step_options.html">Step Specific Arguments</a></li>
							
						</ul>
					</ul>
						<br/>
					<ul>
						<li>Input Files</li>
						<ul>
							
							<li><a href="inputfiles/batchfiles.html">Batch Files</a></li>
							
							
							<li><a href="inputfiles/pedfile.html">PED File</a></li>
							
							
							<li><a href="inputfiles/mapfile.html">MAP File</a></li>
							
							
							<li><a href="inputfiles/tpedfile.html">TPED File</a></li>
							
							
							<li><a href="inputfiles/tfamfile.html">TFAM File</a></li>
							
							
							<li><a href="inputfiles/binary.html">Binary</a></li>
							
							
							<li><a href="inputfiles/covtraitfile.html">Covariates/Traits</a></li>
							

							
							<li><a href="inputfiles/freqfile.html">Frequency File</a></li>
							
							
							<li><a href="inputfiles/strandfile.html">Strand File</a></li>
							
							
							<li><a href="inputfiles/mapfileext.html">Extended Map File</a></li>
							
							
							<li><a href="inputfiles/sampfileext.html">Extended Sample File</a></li>
							
							
							<li><a href="inputfiles/incexclists.html">Include/Exclude Files</a></li>
							
							
							<li><a href="inputfiles/stratfile.html">Stratification File</a></li>
							
							
							<li><a href="inputfiles/filterprocesscfg.html">Filter Process Config.</a></li>
							

						</ul>
					</ul>
						<br/>

					<ul>
						<li>QC Steps</li>
								<ul>
									
										<li>
										<a href="steps/1.html">Marker Genotype Efficiency</a>
										</li>
									
										<li>
										<a href="steps/2.html">Allele Frequency</a>
										</li>
									
										<li>
										<a href="steps/3.html">Concordance</a>
										</li>
									
										<li>
										<a href="steps/4.html">Gender Error</a>
										</li>
									
										<li>
										<a href="steps/5.html">Hardy-Weinberg</a>
										</li>
									
										<li>
										<a href="steps/6.html">Mendelian Errors</a>
										</li>
									
										<li>
										<a href="steps/7.html">Mitochondrial Error</a>
										</li>
									
										<li>
										<a href="steps/8.html">Sample Efficiency</a>
										</li>
									
										<li>
										<a href="steps/33.html">Plato Filter Processing</a>
										</li>
									
								</ul>
						</ul>
						<br/>

						<ul>
							<li>Analysis Steps</li>
								<ul>
									
										<li>
											<a href="steps/9.html">Chisquare</a>
										</li>
										
										<li>
											<a href="steps/10.html">Logistic Regression</a>
										</li>
										
										<li>
											<a href="steps/11.html">Linear Regression</a>
										</li>
										
										<li>
											<a href="steps/12.html">Deletion Detection</a>
										</li>
										
										<li>
											<a href="steps/13.html">Homozygous</a>
										</li>
										
										<li>
											<a href="steps/14.html">Linkage Disequilibrium</a>
										</li>
										
										<li>
											<a href="steps/15.html">TDT</a>
										</li>
										
										<li>
											<a href="steps/30.html">IBS</a>
										</li>
										
										<li>
											<a href="steps/31.html">CMH Analysis</a>
										</li>
										
										<li>
											<a href="steps/32.html">Data Transformations</a>
										</li>
										
										<li>
											<a href="steps/34.html">(EARTH) Multivariate adaptive regression splines</a>
										</li>
										
										<li>
											<a href="steps/35.html">Kinship</a>
										</li>
										
								</ul>
						</ul>
						<br/>
						<ul>
							<li>Formatting Steps</li>
								<ul>
									
										<li>
											<a href="steps/16.html">Beagle</a>
										</li>
										
										<li>
											<a href="steps/17.html">FBAT</a>
										</li>
										
										<li>
											<a href="steps/18.html">GRR</a>
										</li>
										
										<li>
											<a href="steps/19.html">Lapis</a>
										</li>
										
										<li>
											<a href="steps/20.html">MDR</a>
										</li>
										
										<li>
											<a href="steps/21.html">PDT2</a>
										</li>
										
										<li>
											<a href="steps/22.html">PED File</a>
										</li>
										
										<li>
											<a href="steps/23.html">Phase</a>
										</li>
										
										<li>
											<a href="steps/24.html">PowerMarker</a>
										</li>
										
										<li>
											<a href="steps/25.html">QTDT</a>
										</li>
										
										<li>
											<a href="steps/26.html">Structure</a>
										</li>
										
										<li>
											<a href="steps/27.html">Superlink</a>
										</li>
										
										<li>
											<a href="steps/28.html">TPED</a>
										</li>
										
										<li>
											<a href="steps/29.html">Eigenstrat</a>
										</li>
										
									
								</ul>
						</ul>
				<br/>

			</td>
			<td id="content" class="contents">
				
				<p style="color: green"></p>

					<p>Recent technological advances enable the study of hundreds of thousands of human 
	single-nucleotide polymorphisms at the population level. Because strategies for analyzing 
	these data have not kept pace with the laboratory methods that generate the data, it is 
	unlikely that these advances will immediately lead to an improved understanding of the 
	genetic contribution to common human disease and drug response. Currently, no single analytical 
	method allows us to extract all available information from a whole-genome association study. In 
	fact, no single method can be optimal for all datasets, especially when the genetic architecture 
	for diseases can vary substantially, as is certainly the case. Therefore, an integrative platform 
	is needed to accommodate multiple analytical methods for analysis as we learn more about genetic 
	architecture. As a result, we are developing a system for the analysis of genome-wide association data 
	that will incorporate several analytical approaches as filters to allow a scientist to choose whatever 
	analytical methods they wish to apply. PLATO (PLatform for the Analysis, Translation, and Organization 
	of large-scale data) will incorporate a number of filters to select the important SNPs in a genome-wide 
	association study.</p>
	<br/>
    <p>Whole-genome Association Study Pipeline (WASP) has recently been absorbed into PLATO.  WASP was designed to aid in retrieving, evaluating,
    formatting, and analyzing genotypic and clinical data from the latest large-scale genotyping studies.
    WASP implements a battery of quality control procedures to assess the data. Among the currently available
    procedures are the examination of marker and sample genotyping efficiency, allele frequency calculations,
    checks of Mendelian error (if applicable) and gender discrepancies (based on available chromosome X and Y genotypes),
    and tests of Hardy-Weinberg Equilibrium. Additionally, WASP can retrieve and format data for other software programs such as the
    Graphical Representation of Relationships (<a href="http://www.sph.umich.edu/csg/abecasis/GRR/">GRR</a>) program,
    or <a href="http://pritch.bsd.uchicago.edu/structure.html">STRUCTURE</a>, and 
	<!--<?php echo $html->link('others','/formatsteps/outputbeagle'); ?> --> depending on the nature of the samples
    and the depth of examination the user desires to pursue. Beyond the quality control aspect of this application,
    WASP can perform standard tests of association using the Transmission Disequilibrium Test TDT for family-based
    datasets and the chi-square test of association for case-control datasets.
	</p>
	<br/>
	<p>We are also working on a graphical user interface (GUI) data manager for PLATO called PLATO Viewer.  
	This will allow you to do the exact same things as the command line PLATO, however it will be more 
	user friendly by providing point and click batch setups as well as plotting capabilities and data interaction.
	</p>

			</td>
			</tr>
			</table>
		</div>
		<div id="footer">
			&nbsp;
			
					<a href="admin/login.html">Login</a> 
			
			&nbsp;
		</div>
	</div>
</body>
</html>