<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<title>PLATO: PLatform for the Analysis, Translation, and Organization of large-scale data</title>
<link rel="icon" href="https://chgr.mc.vanderbilt.edu/plato/inputfiles/favicon.ico" type="image/icon" />
<link rel="shortcut icon" href="https://chgr.mc.vanderbilt.edu/plato/inputfiles/favicon.ico" type="image/icon" />
<link href="../stylesheets/cake.generic.css%3F1262635684.css" media="screen" rel="stylesheet" type="text/css" />
<link href="../stylesheets/wasp.css%3F1262635684.css" media="screen" rel="stylesheet" type="text/css" />
</head>
<body>
	<a name="top"></a>
	<div id="container">
		<div id="header">
			<h1>
<!--				<a href="/plato/"><img alt="Wasp_icon32x32" src="/plato/images/WASP_icon32x32.PNG?1262635684" /></a> -->
				<a href="../index.html">PLATO:</a> PLatform for the Analysis, Translation, and Organization of large-scale data
			</h1>
		</div>
		<div id="main" class="main">
			<?php //include("side-menu.php");?>
			<table border=0 cellpadding=0>
			<tr>
						<td id="side-menu" class="side-menu">
		       	<br/>
				<ul>
					<li><a href="../index.html">Home</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu/ritchielab/subscriptions">Download</a></li>
					<li><a href="../main/aboutus.html">About Us</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu/ritchielab/index.php">Ritchie Lab</a></li>
					<li><a href="http://chgr.mc.vanderbilt.edu">CHGR</a></li>
				</ul>
<br/>
<!-- Google CSE Search Box Begins  -->
<form method="GET" action="http://searchvu.vanderbilt.edu/search">
    <input type="text" name="q" maxlength="256" value="" />
    <input type="hidden" name="site" value="default_collection" />  
    <input type="hidden" name="client" value="default_frontend" />
    <input type="hidden" name="proxystylesheet" value="default_frontend" />
    <input type="hidden" name="output" value="xml_no_dtd" />   
    <input type="hidden" name="sitesearch" value="https://edam.mc.vanderbilt.edu/wasp" />
    <input type="submit" name="btnG" value="Search" />
</form>
<!-- Google CSE Search Box Ends -->

				<br/>
				<hr/>
				<br/>
<!--				<?php include("side-menu.php");?>-->
				
					<ul>
					
					<li><a href="../main/usage.html">Usage</a></li>
					
						<ul>
							
							<li><a href="../commandlines.html">Commandline Arguments</a></li>
							
							
							<li><a href="../step_options.html">Step Specific Arguments</a></li>
							
						</ul>
					</ul>
						<br/>
					<ul>
						<li>Input Files</li>
						<ul>
							<b>
							<li><a href="batchfiles.html">Batch Files</a></li>
							</b>
							
							<li><a href="pedfile.html">PED File</a></li>
							
							
							<li><a href="mapfile.html">MAP File</a></li>
							
							
							<li><a href="tpedfile.html">TPED File</a></li>
							
							
							<li><a href="tfamfile.html">TFAM File</a></li>
							
							
							<li><a href="binary.html">Binary</a></li>
							
							
							<li><a href="covtraitfile.html">Covariates/Traits</a></li>
							

							
							<li><a href="freqfile.html">Frequency File</a></li>
							
							
							<li><a href="strandfile.html">Strand File</a></li>
							
							
							<li><a href="mapfileext.html">Extended Map File</a></li>
							
							
							<li><a href="sampfileext.html">Extended Sample File</a></li>
							
							
							<li><a href="incexclists.html">Include/Exclude Files</a></li>
							
							
							<li><a href="stratfile.html">Stratification File</a></li>
							
							
							<li><a href="filterprocesscfg.html">Filter Process Config.</a></li>
							

						</ul>
					</ul>
						<br/>

					<ul>
						<li>QC Steps</li>
								<ul>
									
										<li>
										<a href="../steps/1.html">Marker Genotype Efficiency</a>
										</li>
									
										<li>
										<a href="../steps/2.html">Allele Frequency</a>
										</li>
									
										<li>
										<a href="../steps/3.html">Concordance</a>
										</li>
									
										<li>
										<a href="../steps/4.html">Gender Error</a>
										</li>
									
										<li>
										<a href="../steps/5.html">Hardy-Weinberg</a>
										</li>
									
										<li>
										<a href="../steps/6.html">Mendelian Errors</a>
										</li>
									
										<li>
										<a href="../steps/7.html">Mitochondrial Error</a>
										</li>
									
										<li>
										<a href="../steps/8.html">Sample Efficiency</a>
										</li>
									
										<li>
										<a href="../steps/33.html">Plato Filter Processing</a>
										</li>
									
								</ul>
						</ul>
						<br/>

						<ul>
							<li>Analysis Steps</li>
								<ul>
									
										<li>
											<a href="../steps/9.html">Chisquare</a>
										</li>
										
										<li>
											<a href="../steps/10.html">Logistic Regression</a>
										</li>
										
										<li>
											<a href="../steps/11.html">Linear Regression</a>
										</li>
										
										<li>
											<a href="../steps/12.html">Deletion Detection</a>
										</li>
										
										<li>
											<a href="../steps/13.html">Homozygous</a>
										</li>
										
										<li>
											<a href="../steps/14.html">Linkage Disequilibrium</a>
										</li>
										
										<li>
											<a href="../steps/15.html">TDT</a>
										</li>
										
										<li>
											<a href="../steps/30.html">IBS</a>
										</li>
										
										<li>
											<a href="../steps/31.html">CMH Analysis</a>
										</li>
										
										<li>
											<a href="../steps/32.html">Data Transformations</a>
										</li>
										
										<li>
											<a href="../steps/34.html">(EARTH) Multivariate adaptive regression splines</a>
										</li>
										
										<li>
											<a href="../steps/35.html">Kinship</a>
										</li>
										
								</ul>
						</ul>
						<br/>
						<ul>
							<li>Formatting Steps</li>
								<ul>
									
										<li>
											<a href="../steps/16.html">Beagle</a>
										</li>
										
										<li>
											<a href="../steps/17.html">FBAT</a>
										</li>
										
										<li>
											<a href="../steps/18.html">GRR</a>
										</li>
										
										<li>
											<a href="../steps/19.html">Lapis</a>
										</li>
										
										<li>
											<a href="../steps/20.html">MDR</a>
										</li>
										
										<li>
											<a href="../steps/21.html">PDT2</a>
										</li>
										
										<li>
											<a href="../steps/22.html">PED File</a>
										</li>
										
										<li>
											<a href="../steps/23.html">Phase</a>
										</li>
										
										<li>
											<a href="../steps/24.html">PowerMarker</a>
										</li>
										
										<li>
											<a href="../steps/25.html">QTDT</a>
										</li>
										
										<li>
											<a href="../steps/26.html">Structure</a>
										</li>
										
										<li>
											<a href="../steps/27.html">Superlink</a>
										</li>
										
										<li>
											<a href="../steps/28.html">TPED</a>
										</li>
										
										<li>
											<a href="../steps/29.html">Eigenstrat</a>
										</li>
										
										<li>
											<a href="../steps/36.html">Binary (.bim, .bed, .fam) Output</a>
										</li>
										
									
								</ul>
						</ul>
				<br/>

			</td>
			<td id="content" class="contents">
				
				<p style="color: green"></p>

				<h2>Batch Files</h2>
<p>
<font color=red>Note: A batch file is <b>REQUIRED</b> when running PLATO.</font>
<br/>
<br/>
One key feature of PLATO is the use of a "batch file" for quality control and analysis.
This allows the user to specify what QC/Analysis steps to perform in whatever order they desire, including repeating steps and using different trimming thresholds or other criteria.  <br/>
</p>
<br/>
<p>
So, if you want to run Marker Genotype Efficiency, Allele Frequencies, and finally TDT Analysis your batch file would look like:
<pre class="code">
marker-geno-eff
allele-freq
tdt
</pre>
</p>
<br/>
As you can see, each step is listed once per line.  The order of processing would be:
<ol>
    <li><a href="../steps/marker-geno-eff.html">Marker Genotype Efficiency</a></li>
    <li><a href="../steps/allele-freq.html">Allele Frequency</a></li>
    <li><a href="../steps/tdt.html">TDT</a></li>
</ol>
<br/>
By default, all data is processed for each step.  If you would like to remove poor performing samples/markers/families or
restrict processing to certain chromosomes or chromosome locations, you can use a series of <a href="../step_options.html">step specific options</a> on the desired steps.
<br/>
<br/>
For instance, if you would like to remove all markers with a genotype efficiency less than 95%:
<pre class="code">
marker-geno-eff -thresh-markers-min 95
</pre>
This would remove all markers less than 95% from all subsequent steps.
<br/>
<br/>
A more complex example is as follows:
<pre class="code">
marker-geno-eff -thresh-markers-min 0.95
sample-geno-eff -thresh-samples-min 0.95
marker-geno-eff
mendelian-error -thresh-markers-max 15 -thresh-families-max 150 -zero
ped-output
allele-freq -thresh-markers-min 0.05
hw -thresh-markers-min 0.0001
gender-error -chrom 23
tdt
ped-output
grr-output -bp-space 1000000
structure-output -strat-file stratinfo.txt -chrom 15 /
                 -bp-min 5000000 -bp-max 15000000 -bp-space 500000
</pre>
<br/>
As you can see, the use of a batch file makes PLATO highly flexible in that you have the ability to highly customize
your QC and it allows you to take "snapshots" of your data at different stages of the QC/analysis process.  Below is a synopsis of what happens in the above scenario:
<br/>
<br/>
<ol>
    <li style="list-style: decimal">Marker genotype efficiency.  All markers with efficiency less than 95% will be removed from future steps.</li>
    <li style="list-style: decimal">Sample genotype efficiency.  All samples with efficiency less than 95% will be removed from future steps.</li>
    <li style="list-style: decimal">Marker genotype efficiency.  Calculates the marker genotype efficiency again now that all poor performing samples have been removed.</li>
    <li style="list-style: decimal">Mendelian error check.  Removes all markers with more than 15 mendelian errors and families with more than 150 mendelian errors.  It also zeros out the conflicting genotypes.</li>
    <li style="list-style: decimal">PED File output.  Create a PED file based on the "good" markers/samples/genotypes.</li>
    <li style="list-style: decimal">Allele Frequency.  Calculate allele frequency and remove markers with minor allele frequency less than 0.05.</li>
    <li style="list-style: decimal">Hardy-Weinberg.  Calculate Hardy-Weinberg and remove markers with a p-value less than 0.0001.</li>
    <li style="list-style: decimal">Gender Error Check on X chrom.  Find heterozygous males on the X chromosome.  -chrom 23 forces the X chromosome, but is not necessary.</li>
    <li style="list-style: decimal">TDT test.</li>
    <li style="list-style: decimal">PED File output.  Create a second PED file based on the current remaining data set.</li>
    <li style="list-style: decimal">GRR Output.  Create GRR output with a base-pair spacing of 1000000.</li>
    <li style="list-style: decimal">Structure Output.  Create Structure output with stratification information in "stratinfo.txt", use only Chromosome 15 between base-pairs: 5000000-15000000 and use spacing of 500000.</li>
</ol>

			</td>
			</tr>
			</table>
		</div>
		<div id="footer">
			&nbsp;
			
					<a href="../admin/login.html">Login</a> 
			
			&nbsp;
		</div>
	</div>
</body>
</html>